src/scene/shader-lib/chunks/lit/frag/lighting/lightFunctionLight.js

```diff
@@ -0,0 +1,203 @@
+                    #else
+                        vec3 cookieAttenuation = getCookie2DClip(light{i}_cookie, light{i}_shadowMatrix, light{i}_cookieIntensity).{LIGHT{i}COOKIE_CHANNEL};
+                    #endif
+                #endif
+            #endif
+
+            #if LIGHT{i}TYPE == OMNI
+                vec3 cookieAttenuation = getCookieCube(light{i}_cookie, light{i}_shadowMatrix, light{i}_cookieIntensity).{LIGHT{i}COOKIE_CHANNEL};
+            #endif
+
+            // multiply light color by the cookie attenuation
+            lightColor *= cookieAttenuation;
+
+        #endif
+
+        // distance falloff
+        #if LIGHT{i}SHAPE == PUNCTUAL
+            #if LIGHT{i}FALLOFF == LINEAR
+                dAtten = getFalloffLinear(light{i}_radius, lightDirW);
+            #else
+                dAtten = getFalloffInvSquared(light{i}_radius, lightDirW);
+            #endif
+        #else
+            // non punctual lights only gets the range window here
+            dAtten = getFalloffWindow(light{i}_radius, lightDirW);
+        #endif
+
+        if (dAtten < 0.00001) {
+            return;
+        }
+
+        // spot light angle falloff
+        #if LIGHT{i}TYPE == SPOT
+            #if !defined(LIGHT{i}COOKIE) || defined(LIGHT{i}COOKIE_FALLOFF)
+                dAtten *= getSpotEffect(light{i}_direction, light{i}_innerConeAngle, light{i}_outerConeAngle, dLightDirNormW);
+            #endif
+        #endif
+    #endif
+
+    // diffuse lighting - LTC lights do not mix diffuse lighting into attenuation that affects specular
+    #if LIGHT{i}SHAPE != PUNCTUAL
+
+        // attenDiffuse - separate diffuse attenuation for non-punctual light sources
+
+        #if LIGHT{i}TYPE == DIRECTIONAL
+            // NB: A better approximation perhaps using wrap lighting could be implemented here
+            float attenDiffuse = getLightDiffuse(litArgs_worldNormal, dViewDirW, dLightDirNormW);
+        #else
+            // 16.0 is a constant that is in getFalloffInvSquared()
+            #if LIGHT{i}SHAPE == RECT
+                float attenDiffuse = getRectLightDiffuse(litArgs_worldNormal, dViewDirW, lightDirW, dLightDirNormW) * 16.0;
+            #elif LIGHT{i}SHAPE == DISK
+                float attenDiffuse = getDiskLightDiffuse(litArgs_worldNormal, dViewDirW, lightDirW, dLightDirNormW) * 16.0;
+            #elif LIGHT{i}SHAPE == SPHERE
+                float attenDiffuse = getSphereLightDiffuse(litArgs_worldNormal, dViewDirW, lightDirW, dLightDirNormW) * 16.0;
+            #endif
+        #endif
+    #else
+        // one parameter is unused for punctual lights
+        dAtten *= getLightDiffuse(litArgs_worldNormal, vec3(0.0), dLightDirNormW);
+    #endif
+
+    // apply the shadow attenuation
+    #ifdef LIGHT{i}CASTSHADOW
+
+        #if LIGHT{i}TYPE == DIRECTIONAL
+            float shadow = getShadow{i}(vec3(0.0)); //////////////////////////////
```

Comment:
What's with this comment?

Reply:
something I missed, but fixed in follow up PR I have almost ready